File: Algorithms/src/1871. Jump Game VII/Code/solution.py
class Solution:
    def canReach(self, s: str, minJump: int, maxJump: int) -> bool:

        n = len(s)

        # Queue for BFS traversal
        from collections import deque
        q = deque([0])

        # Visited array
        visited = [False] * n
        visited[0] = True

        # Farthest processed position
        far = 0

        while q:

            i = q.popleft()

            # If last index is reached
            if i == n - 1:
                return True

            # Valid jump range
            start = max(i + minJump, far + 1)
            end = min(i + maxJump, n - 1)

            # Explore all possible next positions
            for j in range(start, end + 1):

                # Only move to positions containing '0'
                if s[j] == '0' and not visited[j]:
                    visited[j] = True
                    q.append(j)

            # Update processed boundary
            far = max(far, end)

        return False

File: Algorithms/src/1871. Jump Game VII/Code/solution.cpp
class Solution {
public:
    bool canReach(string s, int minJump, int maxJump) {
        int n = s.size();

        // Queue for BFS traversal
        queue<int> q;

        // Visited array to avoid revisiting indices
        vector<bool> visited(n, false);

        // Start from index 0
        q.push(0);
        visited[0] = true;

        // Keeps track of the farthest processed index
        int far = 0;

        while (!q.empty()) {
            int i = q.front();
            q.pop();

            // If we reached last index, answer is true
            if (i == n - 1)
                return true;

            // Calculate jump range
            int start = max(i + minJump, far + 1);
            int end = min(i + maxJump, n - 1);

            // Explore all valid next positions
            for (int j = start; j <= end; j++) {

                // Only visit positions containing '0'
                if (s[j] == '0' && !visited[j]) {
                    visited[j] = true;
                    q.push(j);
                }
            }

            // Update farthest processed position
            far = max(far, end);
        }

        return false;
    }
};